trl/trainer/dpo_trainer.py

```diff
@@ -815,27 +607,28 @@ def make_inputs_require_grad(module, input, output):
 
             # tokenize the dataset, lower writer batch size to avoid OOM (frequent in vision models)
             fn_kwargs = {
-                "tokenizer": self.processing_class,
-                "args": args,
-                "processor": self.processor if self.is_vision_model else None,
-                "model": model if self.is_encoder_decoder else None,
+                "processing_class": processing_class,
+                "max_prompt_length": args.max_prompt_length,
+                "max_completion_length": args.max_completion_length,
+                # for enc-dec, we add the special tokens ([bos_token] + prompt + [eos_token]; completion + [eos_token])
+                "add_special_tokens": self.is_encoder_decoder,
             }
             train_dataset = train_dataset.map(
-                _tokenize,
+                self.tokenize_row if not self.is_vision_model else self.process_row,
                 fn_kwargs=fn_kwargs,
-                batched=True,
```

Comment:
Does this have any significant impact on preprocessing? I recall @kashif looked into this because the preprocessing was really slow for large preference datasets like Anthropic's HHH

Reply:
In fact, with batch processing, the problem is just shifted, as this function takes a list of examples instead of a single example, but processes it with a for loop.
The advantage of using `batched=False` is that you can make more intensive use of multiprocessing, which should be able to solve the slowness problem.

But I didn't know it was for that reason that we moved to `batched=True` (it wasn't mentioned in #1914). I'll double-check anyway to verify what I said.

Reply:
|          | `dataset_num_proc = None` | `dataset_num_proc = 32` |
| -------- | ------------------------- | ----------------------- |
| `main`   | 04:59                     | 00:38                   |
| PR #2209 | 02:02                     | 00:17                   |

It's worth noting that the comparison isn't entirely fair, because in the case of the 2209, we don't compute the attention mask and the labels in the preprocessing function.

Reply:
this looks great. let's avoid batching and just use multiprocessing as you have